### PR TITLE
Script for ES Snapshots

### PR DESCRIPTION
**Pull request by ranguard**
You can:

- setup
- snap
- list
- restore

Built so that it's easy to put quite complex configs into cron

### CODE REVIEW

File: lib/MetaCPAN/Script/Snapshot.pm [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+}
+
+__PACKAGE__->meta->make_immutable;
+1;
+
+__END__
+
+=head1 NAME
+
+MetaCPAN::Script::Backup - Backup indices and types
```
Comment (mickeyn) [inline lib/MetaCPAN/Script/Snapshot.pm]: you probably want to change the name here :)

File: lib/MetaCPAN/Script/Snapshot.pm [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+use Log::Contextual qw( :log );
+
+use MetaCPAN::Types qw( Bool Int Str File );
+use Moose;
+use DateTime;
+use Try::Tiny;
+use Sys::Hostname;
+use HTTP::Tiny;
+use Cpanel::JSON::XS;
+use DDP;
```
Comment (oalders) [inline lib/MetaCPAN/Script/Snapshot.pm]: Could we alpha-sort these and explicitly not import anything by default? ie `use DateTime ();`

File: lib/MetaCPAN/Script/Snapshot.pm [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+# So we dont' break production
+my $bucket = "mc-${mode}-backups";
+
+my $repository_name = 'our_backups';
+
+## Modes
+has setup => (
+    is            => 'ro',
+    isa           => Bool,
+    documentation => 'Setup the connection with ES',
```
Comment (oalders) [inline lib/MetaCPAN/Script/Snapshot.pm]: Could we set explicit defaults for the `Bool` attributes?  I'm assuming they default to `false`, but I just tried to look it up and the Moose docs weren't clear to me.

File: lib/MetaCPAN/Script/Snapshot.pm [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+has restore => (
+    is            => 'ro',
+    isa           => Bool,
+    documentation => 'Perform a restore',
+);
+
+## Options
+has snap_stub => (
+    is  => 'ro',
+    isa => 'Str',
```
Comment (oalders) [inline lib/MetaCPAN/Script/Snapshot.pm]: You've imported `Str`, so shouldn't be quoted.

File: lib/MetaCPAN/Script/Snapshot.pm [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+    documentation =>
+        'Stub of snapshot name ( e.g full, user etc ), used with dateformat to create the actual name in S3',
+);
+
+has date_format => (
+    is            => 'ro',
+    isa           => 'Str',
```
Comment (oalders) [inline lib/MetaCPAN/Script/Snapshot.pm]: You've imported `Str`, so shouldn't be quoted.

File: lib/MetaCPAN/Script/Snapshot.pm [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+    documentation => 'strftime format to add to snapshot name (eg %Y-%m-%d)',
+);
+
+has snap_name => (
+    is            => 'ro',
+    isa           => 'Str',
```
Comment (oalders) [inline lib/MetaCPAN/Script/Snapshot.pm]: You've imported `Str`, so shouldn't be quoted.

File: lib/MetaCPAN/Script/Snapshot.pm [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+    documentation => 'Full name of snapshot to restore',
+);
+
+has host => (
+    is            => 'ro',
+    isa           => 'Str',
```
Comment (oalders) [inline lib/MetaCPAN/Script/Snapshot.pm]: You've imported `Str`, so shouldn't be quoted.

File: lib/MetaCPAN/Script/Snapshot.pm [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+    default       => 'http://localhost:9200',
+    documentation => 'ES host, defaults to: http://localhost:9200',
+);
+
+# Note: can take wild cards https://www.elastic.co/guide/en/elasticsearch/reference/2.4/multi-index.html
+has indices => (
+    is      => 'ro',
+    isa     => 'ArrayRef',
```
Comment (oalders) [inline lib/MetaCPAN/Script/Snapshot.pm]: We should import `ArrayRef` and use that here.

File: lib/MetaCPAN/Script/Snapshot.pm [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+    $self->date_format || die 'Missing date-format (e.g. %Y-%m-%d)';
+
+    my $date      = DateTime->now->strftime( $self->date_format );
+    my $snap_name = $self->snap_stub . '_' . $date;
+
+    my $indices = join ',', @{ $self->indices };
+    my $data = {
+        "indices"              => $indices,
+        "ignore_unavailable"   => 0,
+        "include_global_state" => 1
```
Comment (oalders) [inline lib/MetaCPAN/Script/Snapshot.pm]: Keys don't need to be quoted.  Missing trailing comma.  Could we alpha sort keys as well?

File: lib/MetaCPAN/Script/Snapshot.pm [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+    };
+
+    log_debug { 'snapping: ' . $snap_name };
+    log_debug { 'with indices: ' . $indices };
+
+    my $path = "${repository_name}/${snap_name}";
+
+    my $response = $self->_request( 'put', $path, $data );
+    return $response;
+
```
Comment (oalders) [inline lib/MetaCPAN/Script/Snapshot.pm]: Extra newline. ;)

File: lib/MetaCPAN/Script/Snapshot.pm [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+    my $data = eval { decode_json $response->{content} };
+
+    foreach my $snapshot ( @{ $data->{snapshots} || [] } ) {
+        log_info { $snapshot->{snapshot} }
+        log_debug { np($snapshot) }
+    }
+
+    return $response;
+
```
Comment (oalders) [inline lib/MetaCPAN/Script/Snapshot.pm]: Extra newline.

File: lib/MetaCPAN/Script/Snapshot.pm [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+}
+
+sub run_list_snaps {
+    my $self = shift;
+
+    my $path = "${repository_name}/_all";
+    my $response = $self->_request( 'get', $path, {} );
+
+    my $data = eval { decode_json $response->{content} };
```
Comment (oalders) [inline lib/MetaCPAN/Script/Snapshot.pm]: We're ok with nothing happening if the JSON doesn't parse?

File: lib/MetaCPAN/Script/Snapshot.pm [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+}
+
+sub run_restore {
+    my $self = shift;
+
+    my $snap_name = $self->snap_name;
+
+    $self->are_you_sure('Restoring... will rename indices to restored_XX');
+
+    # This is a safetly feature, we can always
```
Comment (oalders) [inline lib/MetaCPAN/Script/Snapshot.pm]: *safety

File: lib/MetaCPAN/Script/Snapshot.pm [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+        "type"     => "s3",
+        "settings" => {
+            "bucket"                 => $bucket,
+            "region"                 => "us-east",
+            "protocol"               => "https",
+            "access_key"             => $self->aws_key,
+            "secret_key"             => $self->aws_secret,
+            "server_side_encryption" => 1,
+            "storage_class"          => "standard",
+            "canned_acl"             => "private",
```
Comment (oalders) [inline lib/MetaCPAN/Script/Snapshot.pm]: Sort keys?  No need to quote keys here.

File: lib/MetaCPAN/Script/Snapshot.pm [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+    my $response = $self->_request( 'put', $path, $data );
+    return $response;
+}
+
+sub _request {
+    my ( $self, $method, $path, $data ) = @_;
+
+    my $url = $self->host . '/_snapshot/' . $path;
+
+    my $json = encode_json $data;
```
Comment (oalders) [inline lib/MetaCPAN/Script/Snapshot.pm]: Should probably use brackets unless the function is built in.

File: lib/MetaCPAN/Script/Snapshot.pm [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+
+        log_error { 'Problem requesting ' . $url };
+
+        my $resp_json = eval { decode_json $response->{content} };
+        if ( my $error = $@ ) {
+            log_error { 'Error msg: ' . $response->{content} }
+        }
+        else {
+            log_error { 'Error response: ' . np($resp_json) }
+        }
```
Comment (oalders) [inline lib/MetaCPAN/Script/Snapshot.pm]: Any reason not to use `Try::Tiny` here?

File: lib/MetaCPAN/Script/Snapshot.pm [hunk 17]
```diff
@@ -0,0 +1,289 @@
...
+}
+
+__PACKAGE__->meta->make_immutable;
+1;
+
+__END__
+
+=head1 NAME
+
+MetaCPAN::Script::Snapshot - Snapshot (and restore) ElasticSearch indices
```
Comment (oalders) [inline lib/MetaCPAN/Script/Snapshot.pm]: s/ElasticSearch/Elasticsearch/

**Comment by oalders** [comment]
Don't feel like you have to change everything, but it would be nice to make those changes to the Moose attributes and also use `Try::Tiny` with new code.

File: lib/MetaCPAN/Script/Snapshot.pm (hunk 10 above)
Comment (ranguard) [inline lib/MetaCPAN/Script/Snapshot.pm]: You missed one - there was another extra new line! :)

File: lib/MetaCPAN/Script/Snapshot.pm (hunk 12 above)
Comment (ranguard) [inline lib/MetaCPAN/Script/Snapshot.pm]: Yea, this is on 'list' which you'd only ever run manually in anycase - actual error responses from ES are handled

File: lib/MetaCPAN/Script/Snapshot.pm (hunk 11 above)
Comment (ranguard) [inline lib/MetaCPAN/Script/Snapshot.pm]: Doh - ok - so you found it

**Pull request merged**